File: js/views/new-session.js
import { getPlayers, addSession } from '../store.js';
import { tracks, cars } from '../data.js';

let selectedTrack = null;
let selectedCar = null;
let selectedPlayers = new Set();
let trackSearch = '';
let carSearch = '';

export function render() {
  const players = getPlayers();

  const trackCategories = {};
  tracks.forEach(t => {
    const cat = t.country || 'Other';
    if (!trackCategories[cat]) trackCategories[cat] = [];
    trackCategories[cat].push(t);
  });

  const carCategories = {};
  cars.forEach(c => {
    if (!carCategories[c.category]) carCategories[c.category] = [];
    carCategories[c.category].push(c);
  });

  return `
    <div class="new-session-view">
      <h1 class="section-title">Nueva Carrera</h1>

      <div class="session-form">
        <!-- Date -->
        <div class="form-section">
          <label class="form-label">Fecha</label>
          <input type="date" id="session-date" class="form-input" value="${new Date().toISOString().split('T')[0]}" />
        </div>

        <!-- Track Selection -->
        <div class="form-section">
          <label class="form-label">Pista</label>
          <input type="text" id="track-search" class="form-input search-input" placeholder="Buscar pista..." autocomplete="off" />
          <div class="selection-grid" id="track-grid">
            ${Object.entries(trackCategories).map(([country, trks]) => `
              <div class="selection-category" data-country="${country.toLowerCase()}">
                <div class="category-label">${country}</div>
                ${trks.map(t => `
                  <button type="button" class="selection-item track-item ${selectedTrack === t.id ? 'selected' : ''}" data-id="${t.id}" data-name="${t.name.toLowerCase()}">
                    ${t.name}
                  </button>
                `).join('')}
              </div>
            `).join('')}
          </div>
        </div>

        <!-- Car Selection -->
        <div class="form-section">
          <label class="form-label">Carro</label>
          <input type="text" id="car-search" class="form-input search-input" placeholder="Buscar carro..." autocomplete="off" />
          <div class="selection-grid" id="car-grid">
            ${Object.entries(carCategories).map(([cat, crs]) => `
              <div class="selection-category" data-category="${cat.toLowerCase()}">
                <div class="category-label">${cat}</div>
                ${crs.map(c => `
                  <button type="button" class="selection-item car-item ${selectedCar === c.id ? 'selected' : ''}" data-id="${c.id}" data-name="${c.name.toLowerCase()}">
                    ${c.name}
                  </button>
                `).join('')}
              </div>
            `).join('')}
          </div>
        </div>

        <!-- Player Selection -->
        <div class="form-section">
          <label class="form-label">Pilotos</label>
          ${players.length < 2 ? `
            <div class="form-warning">
              Necesitas al menos 2 pilotos. <a href="#/players">Agregar pilotos</a>
            </div>
          ` : `
            <div class="players-checkboxes">
              ${players.map(p => `
                <label class="player-checkbox ${selectedPlayers.has(p.id) ? 'checked' : ''}">
                  <input type="checkbox" value="${p.id}" ${selectedPlayers.has(p.id) ? 'checked' : ''} />
                  <span>${p.name}</span>
                </label>
              `).join('')}
            </div>
          `}
        </div>

        <!-- Lap Times (shown when players are selected) -->
        <div class="form-section" id="lap-times-section" style="${selectedPlayers.size > 0 ? '' : 'display:none'}">
          <label class="form-label">Tiempos</label>
          <div class="lap-times-list" id="lap-times-list">
            ${Array.from(selectedPlayers).map(pid => {
              const player = players.find(p => p.id === pid);
              if (!player) return '';
              return `
                <div class="lap-time-row">
                  <span class="lap-time-driver">${player.name}</span>
                  <div class="lap-time-inputs">
                    <input type="number" class="time-input time-min" data-player="${pid}" placeholder="0" min="0" max="59" />
                    <span class="time-sep">:</span>
                    <input type="number" class="time-input time-sec" data-player="${pid}" placeholder="00" min="0" max="59" />
                    <span class="time-sep">.</span>
                    <input type="number" class="time-input time-ms" data-player="${pid}" placeholder="000" min="0" max="999" />
                  </div>
                </div>
              `;
            }).join('')}
          </div>
        </div>

        <!-- Submit -->
        <div class="form-actions">
          <button type="button" class="btn btn-primary btn-lg" id="save-session">Guardar Carrera</button>
          <a href="#/sessions" class="btn btn-outline">Cancelar</a>
        </div>
      </div>
    </div>
  `;
}

export function init() {
  // Track search
  const trackSearchInput = document.getElementById('track-search');
  if (trackSearchInput) {
    trackSearchInput.addEventListener('input', (e) => {
      const q = e.target.value.toLowerCase();
      document.querySelectorAll('#track-grid .selection-item').forEach(item => {
        const match = item.dataset.name.includes(q);
        item.style.display = match ? '' : 'none';
      });
      document.querySelectorAll('#track-grid .selection-category').forEach(cat => {
        const visible = cat.querySelectorAll('.selection-item[style=""], .selection-item:not([style])');
        cat.style.display = visible.length > 0 ? '' : 'none';
      });
    });
  }

  // Car search
  const carSearchInput = document.getElementById('car-search');
  if (carSearchInput) {
    carSearchInput.addEventListener('input', (e) => {
      const q = e.target.value.toLowerCase();
      document.querySelectorAll('#car-grid .selection-item').forEach(item => {
        const match = item.dataset.name.includes(q);
        item.style.display = match ? '' : 'none';
      });
      document.querySelectorAll('#car-grid .selection-category').forEach(cat => {
        const visible = cat.querySelectorAll('.selection-item[style=""], .selection-item:not([style])');
        cat.style.display = visible.length > 0 ? '' : 'none';
      });
    });
  }

  // Track selection
  document.querySelectorAll('.track-item').forEach(btn => {
    btn.addEventListener('click', () => {
      document.querySelectorAll('.track-item').forEach(b => b.classList.remove('selected'));
      btn.classList.add('selected');
      selectedTrack = btn.dataset.id;
    });
  });

  // Car selection
  document.querySelectorAll('.car-item').forEach(btn => {
    btn.addEventListener('click', () => {
      document.querySelectorAll('.car-item').forEach(b => b.classList.remove('selected'));
      btn.classList.add('selected');
      selectedCar = btn.dataset.id;
    });
  });

  // Player checkboxes
  document.querySelectorAll('.player-checkbox input').forEach(cb => {
    cb.addEventListener('change', () => {
      if (cb.checked) {
        selectedPlayers.add(cb.value);
      } else {
        selectedPlayers.delete(cb.value);
      }
      cb.closest('.player-checkbox').classList.toggle('checked', cb.checked);
      // Re-render lap times
      window.app.navigate(window.location.hash);
    });
  });

  // Save session
  const saveBtn = document.getElementById('save-session');
  if (saveBtn) {
    saveBtn.addEventListener('click', () => {
      const date = document.getElementById('session-date')?.value;

      if (!selectedTrack) { alert('Selecciona una pista'); return; }
      if (!selectedCar) { alert('Selecciona un carro'); return; }
      if (selectedPlayers.size < 2) { alert('Selecciona al menos 2 pilotos'); return; }

      const results = [];
      let valid = true;

      selectedPlayers.forEach(pid => {
        const min = document.querySelector(`.time-min[data-player="${pid}"]`)?.value || '0';
        const sec = document.querySelector(`.time-sec[data-player="${pid}"]`)?.value || '0';
        const ms = document.querySelector(`.time-ms[data-player="${pid}"]`)?.value || '0';

        const minVal = parseInt(min) || 0;
        const secVal = parseInt(sec) || 0;
        const msVal = parseInt(ms) || 0;

        if (minVal === 0 && secVal === 0 && msVal === 0) {
          valid = false;
          return;
        }

        const lapTime = `${minVal}:${String(secVal).padStart(2, '0')}.${String(msVal).padStart(3, '0')}`;
        results.push({ playerId: pid, lapTime });
      });

      if (!valid) { alert('Ingresa el tiempo de todos los pilotos'); return; }

      const session = addSession({
        date,
        trackId: selectedTrack,
        carId: selectedCar,
        results,
      });

      // Reset state
      selectedTrack = null;
      selectedCar = null;
      selectedPlayers.clear();

      window.location.hash = `#/sessions/${session.id}`;
    });
  }
}

File: volantazo/lib/screens/new_session.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import '../services/storage_service.dart';
import '../models/session.dart';
import '../data/tracks.dart';
import '../data/cars.dart';
import '../theme.dart';

class NewSessionScreen extends StatefulWidget {
  final StorageService storage;

  const NewSessionScreen({super.key, required this.storage});

  @override
  State<NewSessionScreen> createState() => _NewSessionScreenState();
}

class _NewSessionScreenState extends State<NewSessionScreen> {
  String? _selectedTrack;
  String? _selectedCar;
  final Set<String> _selectedPlayers = {};
  final Map<String, List<TextEditingController>> _timeControllers = {};
  final Map<String, bool> _dnfFlags = {};
  DateTime _date = DateTime.now();
  String _trackSearch = '';
  String _carSearch = '';

  @override
  void dispose() {
    for (final controllers in _timeControllers.values) {
      for (final c in controllers) {
        c.dispose();
      }
    }
    super.dispose();
  }

  void _togglePlayer(String id) {
    setState(() {
      if (_selectedPlayers.contains(id)) {
        _selectedPlayers.remove(id);
        _timeControllers[id]?.forEach((c) => c.dispose());
        _timeControllers.remove(id);
        _dnfFlags.remove(id);
      } else {
        _selectedPlayers.add(id);
        _timeControllers[id] = [
          TextEditingController(),
          TextEditingController(),
          TextEditingController(),
        ];
        _dnfFlags[id] = false;
      }
    });
  }

  void _save() {
    if (_selectedTrack == null) {
      _showError('Selecciona una pista');
      return;
    }
    if (_selectedCar == null) {
      _showError('Selecciona un carro');
      return;
    }
    if (_selectedPlayers.length < 2) {
      _showError('Selecciona al menos 2 pilotos');
      return;
    }

    final results = <RaceResult>[];
    for (final pid in _selectedPlayers) {
      final isDnf = _dnfFlags[pid] ?? false;

      if (isDnf) {
        results.add(RaceResult(playerId: pid, lapTime: '0:00.000', dnf: true));
        continue;
      }

      final controllers = _timeControllers[pid]!;
      final min = int.tryParse(controllers[0].text) ?? 0;
      final sec = int.tryParse(controllers[1].text) ?? 0;
      final ms = int.tryParse(controllers[2].text) ?? 0;

      if (min == 0 && sec == 0 && ms == 0) {
        _showError('Ingresa el tiempo de todos los pilotos (o marca DNF)');
        return;
      }

      final lapTime = '$min:${sec.toString().padLeft(2, '0')}.${ms.toString().padLeft(3, '0')}';
      results.add(RaceResult(playerId: pid, lapTime: lapTime));
    }

    final session = widget.storage.addSession(RaceSession(
      id: DateTime.now().millisecondsSinceEpoch.toString(),
      date: '${_date.year}-${_date.month.toString().padLeft(2, '0')}-${_date.day.toString().padLeft(2, '0')}',
      trackId: _selectedTrack!,
      carId: _selectedCar!,
      results: results,
    ));

    Navigator.pop(context);
    Navigator.pushNamed(context, '/session', arguments: session.id);
  }

  void _showError(String msg) {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text(msg), backgroundColor: AppTheme.danger),
    );
  }

  @override
  Widget build(BuildContext context) {
    final players = widget.storage.getPlayers();

    final filteredTracks = allTracks.where((t) =>
        t.name.toLowerCase().contains(_trackSearch.toLowerCase()) ||
        t.country.toLowerCase().contains(_trackSearch.toLowerCase())).toList();

    final filteredCars = allCars.where((c) =>
        c.name.toLowerCase().contains(_carSearch.toLowerCase()) ||
        c.category.toLowerCase().contains(_carSearch.toLowerCase())).toList();

    final tracksByCountry = <String, List<Track>>{};
    for (final t in filteredTracks) {
      tracksByCountry.putIfAbsent(t.country, () => []).add(t);
    }

    final carsByCategory = <String, List<Car>>{};
    for (final c in filteredCars) {
      carsByCategory.putIfAbsent(c.category, () => []).add(c);
    }

    return Scaffold(
      appBar: AppBar(
        title: const Text('NUEVA CARRERA'),
        bottom: PreferredSize(
          preferredSize: const Size.fromHeight(3),
          child: Container(height: 3, color: AppTheme.primary),
        ),
      ),
      body: ListView(
        padding: const EdgeInsets.all(16),
        children: [
          // Date
          _label('FECHA'),
          InkWell(
            onTap: () async {
              final picked = await showDatePicker(
                context: context,
                initialDate: _date,
                firstDate: DateTime(2020),
                lastDate: DateTime.now(),
              );
              if (picked != null) setState(() => _date = picked);
            },
            child: Container(
              padding: const EdgeInsets.all(14),
              decoration: BoxDecoration(
                color: AppTheme.surface,
                border: Border.all(color: AppTheme.border),
                borderRadius: BorderRadius.circular(8),
              ),
              child: Row(
                children: [
                  Text('${_date.day}/${_date.month}/${_date.year}'),
                  const Spacer(),
                  const Icon(Icons.calendar_today, size: 18, color: AppTheme.textSecondary),
                ],
              ),
            ),
          ),
          const SizedBox(height: 24),

          // Track
          _label('PISTA'),
          TextField(
            decoration: const InputDecoration(hintText: 'Buscar pista...', prefixIcon: Icon(Icons.search, size: 20)),
            onChanged: (v) => setState(() => _trackSearch = v),
          ),
          const SizedBox(height: 8),
          Container(
            constraints: const BoxConstraints(maxHeight: 220),
            decoration: BoxDecoration(
              color: AppTheme.surface,
              border: Border.all(color: AppTheme.border),
              borderRadius: BorderRadius.circular(8),
            ),
            child: ListView(
              padding: const EdgeInsets.all(8),
              shrinkWrap: true,
              children: tracksByCountry.entries.map((entry) => Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Padding(
                    padding: const EdgeInsets.only(top: 4, bottom: 4, left: 4),
                    child: Text(entry.key, style: const TextStyle(color: AppTheme.textSecondary, fontSize: 11, fontWeight: FontWeight.w600, letterSpacing: 1)),
                  ),
                  Wrap(
                    spacing: 6,
                    runSpacing: 6,
                    children: entry.value.map((t) => ChoiceChip(
                      label: Text(t.name, style: const TextStyle(fontSize: 13)),
                      selected: _selectedTrack == t.id,
                      onSelected: (_) => setState(() => _selectedTrack = t.id),
                    )).toList(),
                  ),
                  const SizedBox(height: 8),
                ],
              )).toList(),
            ),
          ),
          const SizedBox(height: 24),

          // Car
          _label('CARRO'),
          TextField(
            decoration: const InputDecoration(hintText: 'Buscar carro...', prefixIcon: Icon(Icons.search, size: 20)),
            onChanged: (v) => setState(() => _carSearch = v),
          ),
          const SizedBox(height: 8),
          Container(
            constraints: const BoxConstraints(maxHeight: 220),
            decoration: BoxDecoration(
              color: AppTheme.surface,
              border: Border.all(color: AppTheme.border),
              borderRadius: BorderRadius.circular(8),
            ),
            child: ListView(
              padding: const EdgeInsets.all(8),
              shrinkWrap: true,
              children: carsByCategory.entries.map((entry) => Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Padding(
                    padding: const EdgeInsets.only(top: 4, bottom: 4, left: 4),
                    child: Text(entry.key, style: const TextStyle(color: AppTheme.textSecondary, fontSize: 11, fontWeight: FontWeight.w600, letterSpacing: 1)),
                  ),
                  Wrap(
                    spacing: 6,
                    runSpacing: 6,
                    children: entry.value.map((c) => ChoiceChip(
                      label: Text(c.name, style: const TextStyle(fontSize: 13)),
                      selected: _selectedCar == c.id,
                      onSelected: (_) => setState(() => _selectedCar = c.id),
                    )).toList(),
                  ),
                  const SizedBox(height: 8),
                ],
              )).toList(),
            ),
          ),
          const SizedBox(height: 24),

          // Players
          _label('PILOTOS'),
          if (players.length < 2)
            Card(
              child: Padding(
                padding: const EdgeInsets.all(16),
                child: Text('Necesitas al menos 2 pilotos. Ve a la pestaña Pilotos.',
                    style: TextStyle(color: AppTheme.textSecondary)),
              ),
            )
          else
            Wrap(
              spacing: 8,
              runSpacing: 8,
              children: players.map((p) => FilterChip(
                label: Text(p.name),
                selected: _selectedPlayers.contains(p.id),
                onSelected: (_) => _togglePlayer(p.id),
                selectedColor: AppTheme.primary,
                checkmarkColor: Colors.white,
              )).toList(),
            ),
          const SizedBox(height: 24),

          // Lap times
          if (_selectedPlayers.isNotEmpty) ...[
            _label('TIEMPOS'),
            ..._selectedPlayers.map((pid) {
              final player = widget.storage.getPlayerById(pid);
              if (player == null) return const SizedBox.shrink();
              final controllers = _timeControllers[pid]!;
              final isDnf = _dnfFlags[pid] ?? false;
              return Card(
                margin: const EdgeInsets.only(bottom: 8),
                child: Padding(
                  padding: const EdgeInsets.all(12),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Row(
                        children: [
                          Text(player.name, style: const TextStyle(fontWeight: FontWeight.w700)),
                          const Spacer(),
                          FilterChip(
                            label: const Text('DNF', style: TextStyle(fontSize: 12)),
                            selected: isDnf,
                            onSelected: (v) => setState(() => _dnfFlags[pid] = v),
                            selectedColor: AppTheme.danger,
                            checkmarkColor: Colors.white,
                          ),
                        ],
                      ),
                      if (!isDnf) ...[
                        const SizedBox(height: 8),
                        Row(
                          children: [
                            _timeField(controllers[0], 'min', 59),
                            const Padding(padding: EdgeInsets.symmetric(horizontal: 4), child: Text(':', style: TextStyle(fontSize: 20, fontWeight: FontWeight.w700))),
                            _timeField(controllers[1], 'seg', 59),
                            const Padding(padding: EdgeInsets.symmetric(horizontal: 4), child: Text('.', style: TextStyle(fontSize: 20, fontWeight: FontWeight.w700))),
                            _timeField(controllers[2], 'ms', 999, width: 65),
                          ],
                        ),
                      ],
                    ],
                  ),
                ),
              );
            }),
          ],
          const SizedBox(height: 16),

          ElevatedButton(
            onPressed: _save,
            style: ElevatedButton.styleFrom(padding: const EdgeInsets.symmetric(vertical: 16)),
            child: const Text('GUARDAR CARRERA', style: TextStyle(fontSize: 16)),
          ),
          const SizedBox(height: 32),
        ],
      ),
    );
  }

  Widget _label(String text) {
    return Padding(
      padding: const EdgeInsets.only(bottom: 8),
      child: Text(text, style: const TextStyle(color: AppTheme.textSecondary, fontSize: 12, fontWeight: FontWeight.w600, letterSpacing: 1)),
    );
  }

  Widget _timeField(TextEditingController controller, String hint, int max, {double width = 55}) {
    return SizedBox(
      width: width,
      child: TextField(
        controller: controller,
        keyboardType: TextInputType.number,
        textAlign: TextAlign.center,
        style: const TextStyle(fontFamily: 'monospace', fontSize: 16),
        inputFormatters: [FilteringTextInputFormatter.digitsOnly],
        decoration: InputDecoration(
          hintText: hint,
          contentPadding: const EdgeInsets.symmetric(vertical: 12, horizontal: 8),
        ),
      ),
    );
  }
}
